### PR TITLE
Add Qwen2MoE

### PR DESCRIPTION
**Pull request by bozheng-hit**
# Adding Qwen2MoE
This PR adds the support of codes for the coming Qwen2MoE models. For information about Qwen, please visit https://github.com/QwenLM/Qwen. @ArthurZucker

### CODE REVIEW

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 1]
```diff
@@ -0,0 +1,1618 @@
...
+_CONFIG_FOR_DOC = "Qwen2MoeConfig"
+
+QWEN2MOE_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "Qwen/Qwen1.5-MoE-A2.7B",
+    # See all Qwen2 models at https://huggingface.co/models?filter=qwen2
+]
+
+
+# Copied from transformers.models.mixtral.modeling_mixtral.load_balancing_loss_func
+def load_balancing_loss_func(
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:68]: is that also the loss you used for training your MoE? 🤗

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 2]
```diff
@@ -0,0 +1,1618 @@
...
+
+        if config.shared_expert_intermediate_size is not None and config.shared_expert_intermediate_size > 0:
+            self.shared_expert = Qwen2MoeMLP(config, intermediate_size=config.shared_expert_intermediate_size)
+        else:
+            self.shared_expert = None
+
+        if config.shared_expert_gate:
+            self.shared_expert_gate = torch.nn.Linear(config.hidden_size, 1, bias=False)
+        else:
+            self.shared_expert_gate = None
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:826]: in the trained checkpoint what is the value of `shared_expert_gate` or `shared_expert_intermediate_size` . If they are always the same then we don't need if elses

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 3]
```diff
@@ -0,0 +1,1618 @@
...
+
+            # However `index_add_` only support torch tensors for indexing so we'll use
+            # the `top_x` tensor here.
+            final_hidden_states.index_add_(0, top_x, current_hidden_states.to(hidden_states.dtype))
+
+        if self.shared_expert is not None:
+            y = self.shared_expert(hidden_states)
+            if self.shared_expert_gate is not None:
+                y = F.sigmoid(self.shared_expert_gate(hidden_states)) * y
+            final_hidden_states = final_hidden_states + y
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:876]: let's avoid single letter variables

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 4]
```diff
@@ -0,0 +1,1618 @@
...
+            logger.warning_once(
+                f"Sliding Window Attention is enabled but not implemented for `{config._attn_implementation}`; "
+                "unexpected results may be encountered."
+            )
+        self.self_attn = QWEN2MOE_ATTENTION_CLASSES[config._attn_implementation](config, layer_idx)
+
+        if config.num_experts and config.expert_interval is not None and (layer_idx + 1) % config.expert_interval == 0:
+            self.mlp = Qwen2MoeSparseMoeBlock(config)
+        else:
+            self.mlp = Qwen2MoeMLP(config)
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:897]: same here, is it full Moe or `expert_interval != 0`

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 5]
```diff
@@ -0,0 +1,1618 @@
...
+# Modified from transformers.models.mistral.modeling_mistral.MistralMLP with Mistral->Qwen2Moe
+class Qwen2MoeMLP(nn.Module):
+    def __init__(self, config, intermediate_size=None):
+        super().__init__()
+        self.config = config
+        self.hidden_size = config.hidden_size
+        if intermediate_size is None:
+            self.intermediate_size = config.intermediate_size
+        else:
+            self.intermediate_size = intermediate_size
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:258]: you seem to always pass the `intermediate_size`
```suggestion
        self.intermediate_size = intermediate_size
```

**Review by ArthurZucker — COMMENTED** [review]
Thanks a few small comments for a small review for now!

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py (hunk 1 above)
Comment (bozheng-hit) [reply src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:68]: Yes.

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py (hunk 2 above)
Comment (bozheng-hit) [reply src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:826]: This is for checkpoints without shared_expert_gate, but for now there are always a gate in our trained checkpoint.

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py (hunk 5 above)
Comment (bozheng-hit) [reply src/transformers/models/qwen2_moe/modeling_qwen2_moe.py]: The code is updated.

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 6]
```diff
@@ -0,0 +1,1615 @@
...
+
+        if config.shared_expert_intermediate_size is not None and config.shared_expert_intermediate_size > 0:
+            self.shared_expert = Qwen2MoeMLP(config, intermediate_size=config.shared_expert_intermediate_size)
+        else:
+            self.shared_expert = None
+
+        if config.shared_expert_gate:
+            self.shared_expert_gate = torch.nn.Linear(config.hidden_size, 1, bias=False)
+        else:
+            self.shared_expert_gate = None
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:823]: For both of these if / else: 
> for now there are always a gate in our trained checkpoint.

it is `transformers` philosophy to only merge what already has checkpoints. Unless you plan / are training one with this. But that would be a different architecture. 

We try to avoid if else base on the config like this!

**Review by ArthurZucker — COMMENTED** [review]
Thanks! Overall alright, I think one thing we need to be careful about is to test in the `terst_modeling_qwen2_moe` when there are sparse and non sparse. The loss needs to ignore None router logits!

**Comment by JustinLin610** [comment]
Hey, I guess our update of the testing solves the issues mentioned. Take another look?

File: README.md [hunk 7]
```diff
@@ -468,6 +468,7 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
...
 1. **[Qwen2](https://huggingface.co/docs/transformers/model_doc/qwen2)** (from the Qwen team, Alibaba Group) released with the paper [Qwen Technical Report](https://arxiv.org/abs/2309.16609) by Jinze Bai, Shuai Bai, Yunfei Chu, Zeyu Cui, Kai Dang, Xiaodong Deng, Yang Fan, Wenbin Ge, Yu Han, Fei Huang, Binyuan Hui, Luo Ji, Mei Li, Junyang Lin, Runji Lin, Dayiheng Liu, Gao Liu, Chengqiang Lu, Keming Lu, Jianxin Ma, Rui Men, Xingzhang Ren, Xuancheng Ren, Chuanqi Tan, Sinan Tan, Jianhong Tu, Peng Wang, Shijie Wang, Wei Wang, Shengguang Wu, Benfeng Xu, Jin Xu, An Yang, Hao Yang, Jian Yang, Shusheng Yang, Yang Yao, Bowen Yu, Hongyi Yuan, Zheng Yuan, Jianwei Zhang, Xingxuan Zhang, Yichang Zhang, Zhenru Zhang, Chang Zhou, Jingren Zhou, Xiaohuan Zhou and Tianhang Zhu.
+1. **[Qwen2MoE](https://huggingface.co/docs/transformers/main/model_doc/qwen2_moe)** (from the Qwen team, Alibaba Group) released with the paper [blog post](https://qwenlm.github.io/blog/qwen1.5/) by Bo Zheng, Dayiheng Liu, Rui Men, Junyang Lin, Zhou San, Bowen Yu, An Yang, Mingfeng Xue, Fei Huang, Binyuan Hui, Mei Li, Tianyu Liu, Xingzhang Ren, Xuancheng Ren, Kexin Yang, Chang Zhou, Jingren Zhou.
```
Comment (ArthurZucker) [inline README.md:471]:
```suggestion
1. **[Qwen2MoE](https://huggingface.co/docs/transformers/main/model_doc/qwen2_moe)** (from the Qwen team, Alibaba Group) released with [blog post](https://qwenlm.github.io/blog/qwen1.5/) by Bo Zheng, Dayiheng Liu, Rui Men, Junyang Lin, Zhou San, Bowen Yu, An Yang, Mingfeng Xue, Fei Huang, Binyuan Hui, Mei Li, Tianyu Liu, Xingzhang Ren, Xuancheng Ren, Kexin Yang, Chang Zhou, Jingren Zhou.
```

File: docs/source/en/model_doc/qwen2_moe.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+## Overview
+
+Qwen2MoE is the new model series of large language models from the Qwen team. Previously, we released the Qwen series, including Qwen-72B, Qwen-1.8B, Qwen-VL, Qwen-Audio, etc.
+
+### Model Details
+
+Qwen2MoE is a language model series including decoder language models of different model sizes. For each size, we release the base language model and the aligned chat model. Qwen2MoE has the following architectural choices:
+
+- Qwen2MoE is based on the Transformer architecture with SwiGLU activation, attention QKV bias, group query attention, mixture of sliding window attention and full attention, etc. Additionally, we have an improved tokenizer adaptive to multiple natural languages and codes.
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/qwen2_moe.md:27]: I think diff with Llama is more relevant but alright!
Also tokenizer, details are welcome as to what is it that makes it adaptive?

File: docs/source/en/model_doc/qwen2_moe.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+- Qwen2MoE employs Mixture of Experts (MoE) architecture, where the models are upcycled from dense language models. For instance, `Qwen1.5-MoE-A2.7B` is upcycled from `Qwen-1.8B`. It has 14.3B parameters in total and 2.7B activated parameters during runtime, while it achieves comparable performance with `Qwen1.5-7B`, with only 20% of the training resources.
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/qwen2_moe.md:28]: Does upcycle mean that you init all Experts from the dense MoE?

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 10]
```diff
@@ -0,0 +1,1605 @@
...
+        self.hidden_size = config.hidden_size
+
+        if config.use_sliding_window and config._attn_implementation != "flash_attention_2":
+            logger.warning_once(
+                f"Sliding Window Attention is enabled but not implemented for `{config._attn_implementation}`; "
+                "unexpected results may be encountered."
+            )
+        self.self_attn = QWEN2MOE_ATTENTION_CLASSES[config._attn_implementation](config, layer_idx)
+
+        if config.num_experts > 0 and (layer_idx + 1) % config.expert_interval == 0:
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:881]:
```suggestion
        if config.num_experts > 0 and (layer_idx + 1) % config.decoder_sparse_step == 0:
```
we should be unifying the API with `switch_transformers`

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 11]
```diff
@@ -0,0 +1,1605 @@
...
+
+class Qwen2MoeDecoderLayer(nn.Module):
+    def __init__(self, config: Qwen2MoeConfig, layer_idx: int):
+        super().__init__()
+        self.hidden_size = config.hidden_size
+
+        if config.use_sliding_window and config._attn_implementation != "flash_attention_2":
+            logger.warning_once(
+                f"Sliding Window Attention is enabled but not implemented for `{config._attn_implementation}`; "
+                "unexpected results may be encountered."
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:877]: sdpa would support sliding window with correct torch version

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 12]
```diff
@@ -0,0 +1,1605 @@
...
+
+        # Fully Connected
+        residual = hidden_states
+        hidden_states = self.post_attention_layernorm(hidden_states)
+
+        if isinstance(self.mlp, Qwen2MoeSparseMoeBlock):
+            hidden_states, router_logits = self.mlp(hidden_states)
+        else:
+            hidden_states = self.mlp(hidden_states)
+            router_logits = None
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:945]:
```suggestion
        hidden_states =  self.mlp(hidden_states)
        if isinstance(hidden_states, tuple):
            hidden_states, router_logits = hidden_states
        else:
            router_logits = None
```

Comment (ArthurZucker) [reply src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:945]: it's a nit

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 13]
```diff
@@ -0,0 +1,1605 @@
...
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
+
+
+@add_start_docstrings(
+    "The bare Qwen2MoE Model outputting raw hidden-states without any specific head on top.",
+    QWEN2MOE_START_DOCSTRING,
+)
+class Qwen2MoeModel(Qwen2MoePreTrainedModel):
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:1083]: I'd recommend to check the `gemma` model as the model can be simplified + support torch compile if we follow it, but otherwise alright

**Review by ArthurZucker — COMMENTED** [review]
LGTM Almost not comment, I'll do another full review tomorrow but it is pretty much mergable ! 🔥 

Mostly using the latest api would be nice for cache_positions in the Model, but it's really optional!

File: docs/source/en/model_doc/qwen2_moe.md (hunk 9 above)
Comment (bozheng-hit) [reply docs/source/en/model_doc/qwen2_moe.md:28]: Yes, all experts are initialized from the dense model.

File: docs/source/en/model_doc/qwen2_moe.md (hunk 8 above)
Comment (bozheng-hit) [reply docs/source/en/model_doc/qwen2_moe.md:27]: We use the same tokenizer with the Qwen2 models.

File: src/transformers/models/qwen2_moe/configuration_qwen2_moe.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+        self.intermediate_size = intermediate_size
+        self.num_hidden_layers = num_hidden_layers
+        self.num_attention_heads = num_attention_heads
+        self.use_sliding_window = use_sliding_window
+        self.sliding_window = sliding_window
+        self.max_window_layers = max_window_layers
+
+        # for backward compatibility
+        if num_key_value_heads is None:
+            num_key_value_heads = num_attention_heads
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/configuration_qwen2_moe.py:156]:
```suggestion
```
no BC needed

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 15]
```diff
@@ -0,0 +1,1600 @@
...
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "Qwen/Qwen1.5-MoE-A2.7B"
+_CONFIG_FOR_DOC = "Qwen2MoeConfig"
+
+QWEN2MOE_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "Qwen/Qwen1.5-MoE-A2.7B",
+    # See all Qwen2 models at https://huggingface.co/models?filter=qwen2
+]
+
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:63]:
```suggestion
```

File: src/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 16]
```diff
@@ -0,0 +1,1600 @@
...
+    Since it does classification on the last token, it requires to know the position of the last token. If a
+    `pad_token_id` is defined in the configuration, it finds the last token that is not a padding token in each row. If
+    no `pad_token_id` is defined, it simply takes the last value in each row of the batch. Since it cannot guess the
+    padding tokens when `inputs_embeds` are passed instead of `input_ids`, it does the same (take the last value in
+    each row of the batch).
+    """,
+    QWEN2MOE_START_DOCSTRING,
+)
+class Qwen2MoeForSequenceClassification(Qwen2MoePreTrainedModel):
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/modeling_qwen2_moe.py:1495]: missing copied from here

File: src/transformers/models/qwen2_moe/configuration_qwen2_moe.py [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+
+from ...configuration_utils import PretrainedConfig
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+QWEN2MOE_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "Qwen/Qwen1.5-MoE-A2.7B": "https://huggingface.co/Qwen/Qwen1.5-MoE-A2.7B/resolve/main/config.json",
+}
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_moe/configuration_qwen2_moe.py:25]:
```suggestion
```

**Review by ArthurZucker — APPROVED** [review]
LGTM! A few small nits and should be good to go

File: src/transformers/models/qwen2_moe/configuration_qwen2_moe.py (hunk 17 above)
Comment (bozheng-hit) [reply src/transformers/models/qwen2_moe/configuration_qwen2_moe.py]: I cannot pass `make fixup` after removing these lines.

**Comment by bozheng-hit** [comment]
@ArthurZucker Hi, I have rebased our branch and solved all the conflicts. I think the codes are ready to be merged now.

**Comment by ydshieh** [comment]
Hi @bozheng-hit 

Thank you for adding this model.

I see `Qwen/Qwen1.5-MoE-A2.7B` is used in the tests, but I could not find it on the Hub. See 

https://huggingface.co/models?search=Qwen1.5-MoE

Could you check this, make some necessary updates, and make sure the integration tests pass by running

> RUN_SLOW=1 TF_FORCE_GPU_ALLOW_GROWTH=yes python3 -m pytest -v tests/models/qwen2_moe

**Comment by lucasjinreal** [comment]
Didn't saw `Qwen/Qwen1.5-MoE-A2.7B` anywhere, time estimation on new model release?
